[GH-26678] Convert katex class component to function component

marlenekoh opened the pull request:
<!-- Thank you for contributing a pull request! Here are a few tips to help you:

1. If this is your first contribution, make sure you've read the Contribution Checklist https://developers.mattermost.com/contribute/getting-started/contribution-checklist/
2. Read our blog post about "Submitting Great PRs" https://developers.mattermost.com/blog/2019-01-24-submitting-great-prs
3. Take a look at other repository specific documentation at https://developers.mattermost.com/contribute
-->

#### Summary
<!--
A description of what this pull request does, as well as QA test steps (if applicable and if not already added to the Jira ticket).
-->

Migrates `./components/latex_block/latex_block.tsx` from Class Component to Function Component

#### Ticket Link
<!--
If applicable, please include both or either of the following links:

Fixes https://github.com/mattermost/mattermost/issues/XXX
Jira https://mattermost.atlassian.net/browse/MM-XXX
-->

Fixes https://github.com/mattermost/mattermost/issues/26678

#### Screenshots
<!--
If the PR includes UI changes, include screenshots/GIFs.

For an easier comparison of UI changes a table (template below) can be used.

|  before  |  after  |
|----|----|
| <insert before screenshot here> | <insert after screenshot here> |

-->

#### Release Note
<!--
Add a release note for each of the following conditions:

* Config changes (additions, deletions, updates).
* API additions—new endpoint, new response fields, or newly accepted request parameters.
* Database changes (any).
* Schema migration changes. Use the [Schema Migration Template](https://docs.google.com/document/d/18lD7N32oyMtYjFrJKwsNv8yn6Fe5QtF-eMm8nn0O8tk/edit?usp=sharing) as a starting point to capture these details as release notes. 
* Websocket additions or changes.
* Anything noteworthy to a Mattermost instance administrator (err on the side of over-communicating).
* New features and improvements, including behavioral changes, UI changes, and CLI changes.
* Bug fixes and fixes of previous known issues.
* Deprecation warnings, breaking changes, or compatibility notes.

If no release notes are required, write NONE. Use past-tense. Newlines are stripped.

Examples:

```
Added new API endpoints POST /api/v4/foo, GET api/v4/foo, and GET api/v4/foo/:foo_id.
```

```
Added a new config setting ServiceSettings.FooBar. Added a new column Foo to the Users table.
```

```
NONE
```
-->
```release-note

```

mattermost-build commented:
Hello @marlenekoh,

Thanks for your pull request! A Core Committer will review your pull request soon. For code contributions, you can learn more about the review process [here](https://developers.mattermost.com/contribute/getting-started/code-review/).

larkox commented on webapp/channels/src/components/latex_block/latex_block.tsx:
```diff
@@ -2,74 +2,64 @@
-type State = {
-    katex?: Katex;
-}
+const LatexBlock = ({content, enableLatex}: Props) => {
```
Let's separate each prop in a different line. In the future, it helps to verify when we are adding/removing props.
```suggestion
const LatexBlock = ({
    content,
    enableLatex,
}: Props) => {
```

larkox reviewed (changes requested):
Great work! Just one minor thing.

marlenekoh commented:
@larkox thanks for the review!

larkox reviewed (approved):
Great work. Thanks!

hmhealey reviewed (changes requested):
Thanks for the PR, @marlenekoh!

The code looks good to me, but it looks like the unit tests for this component are broken now. Could you take a look at fixing those? You can run just the test for that one component by running `npm test -w channels -- components/latex_block` in the `webapp` folder.

It looks like one of the snapshots is broken because of the component's display name changed, so that one just needs to be updated with `-u`, but the others seem to be failing because it's not loading Katex any more. I'm not sure if that's because `useEffect` isn't setting the state correctly or if the way that we're waiting for the async import to happen doesn't work with the functional component though

marlenekoh commented:
Alright, I will take a look!

marlenekoh commented:
@hmhealey the tests for latex component should pass now, please take a look 🙏 

- I had to change the code from `shallow()` to `render()` in the latex tests because the `useEffect()` hook was not running for `shallow()`
- also added `data-testid` to make explicit which `<div>` block was being returned from the component
- using `await screen.findAllByTestId()` in the tests to remove this error
   ```
    Unexpected console logsWarning: An update to %s inside a test was not wrapped in act(...).

    When testing, code that causes React state updates should be wrapped into act(...):

    act(() => {
      /* fire events that update state */
    });
    /* assert on the output */

    This ensures that you're testing the behavior the user would see in the browser. Learn more at https://reactjs.org/link/wrap-tests-with-act%s,LatexBlock,
   ```
- also updated the snapshot for `message_html_to_component.test.tsx`.

hmhealey reviewed (approved):
Thanks for all the details on what you changed. That makes sense that we'd have to switch to Testing Library for this since Enzyme doesn't even officially support React 17. We really need to get rid of Enzyme one of these days.

As for `throwOnError`, I didn't realize that that's been false for years at this point since it still seems to render errors like it did originally. I see you reverted that though which is also totally fine. It's definitely something we'll want to look into at some point, but it's not at all a priority.

Overall, this looks good to me!

yasserfaraazkhan commented:
/e2e-tests

yasserfaraazkhan reviewed (approved):
<img width="500" alt="image" src="https://github.com/mattermost/mattermost/assets/15979783/d42415be-f8b7-4b7b-b697-0cb1f066ee6a">
